### DISCUSSION TITLE
SonixGPS: How to retrieve / save TrackerDevice data without "Reference" and "Stylus"

### DISCUSSION THREAD

**Discussion by SenonETS** [PlusToolkit/PlusLib#818] [Q&A] [answered]
Hello,

This is a newbie on SonixTablet with Ascension3DG SonixGPS.
Here is the actual [connection photo for the Ascension driveBAY](https://ibb.co/JzZHNc3).

There is only one L14-5 transducer connected. 
I am trying to retrieve and save live ultrasound images / sequences using Slicer and PLUS on a Windows10 64bit computer.

Because only PortName="0" is connected and is able to provide data (while the other two "Port"s are not connected, not sure where are them), I have to comment out everything related to  `Id="Reference" PortName="1"` and `Id="Stylus" PortName="2"` from the official tutorial configuration file.

Please find the [configuration file here](https://file.io/eUHSGBgwto4Y).
It seems that this file link can only be downloaded once. I do not know how to upload files from here (discussion post).
Or if you view it from [here](https://justpaste.it/8dmjs) without downloading it.

It seems that in this way, the only data I can get is still the Type="SonixVideo", with no data related to Type="Ascension3DG".

Could you please share ideas about how to get the live position / transform data corresponding to the Type="SonixVideo" ultrasound image?

**Comment by Sunderlandkyl** [comment, marked as answer]
````
<!-- <Message Type="TRANSFORM" /> -->
````
This needs to be uncommented.

````
<!-- <TransformNames>  -->
  <!-- <Transform Name="ProbeToReference" /> -->
  <!-- <Transform Name="StylusToReference" /> -->
  <!-- <Transform Name="ReferenceToTracker" /> -->
<!-- </TransformNames> -->
````
You need to uncomment TransformNames element, and all of the Transform elements that you want to send in PlusOpenIGTLinkServer.

**Reply by SenonETS** [reply]
I cannot, because "Reference" is not provided; as I said above,  I have to comment out everything related to `Id="Reference" PortName="1"` from configuration file, other wise, you will see the error:

`060321_174319.452|ERROR|046.922000| Transform path not found from Probe to Reference coordinate system. Available transforms in the repository (including the inverse of these transforms): ImageToTransducer (valid, non-persistent), ProbeToTracker (valid, non-persistent), TransducerToProbe (valid, persistent)| in ..\..\..\PlusLib\src\PlusCommon\vtkTransformRepository.cxx(535)`
`060321_174319.452|ERROR|046.922000| Transform path not found from Reference to Tracker coordinate system. Available transforms in the repository (including the inverse of these transforms): ImageToTransducer (valid, non-persistent), ProbeToTracker (valid, non-persistent), TransducerToProbe (valid, persistent)| in ..\..\..\PlusLib\src\PlusCommon\vtkTransformRepository.cxx(535)
`

Here is the [error screenshot](https://ibb.co/BfRpPhb) as you suggested, which is not a solution.

**Reply by Sunderlandkyl** [reply]
It should work if you change ProbeToReference to ProbeToTracker, since Tracker is the ToolReferenceFrame.

**Reply by SenonETS** [reply]
I just did, as you suggested using "ProbeToTracker", [see the code screenshot](https://ibb.co/yXzGB7p)

The error does not change, exactly the same with [the above one](https://ibb.co/BfRpPhb).

To make things clear, here is another screenshot of the [whole code](https://ibb.co/FmK5brb), the whole code as you suggested that generates error above.

**Reply by Sunderlandkyl** [reply]
StylusToReference and ReferenceToTracker are not valid transforms since the Stylus and Reference tools are not connected.

Can you try changing the TransformNames element so that it contains only the one transform?
````
<TransformNames>
  <Transform Name="ProbeToTracker" />
</TransformNames>
````

**Reply by SenonETS** [reply]
Thank you, Sunderlankyl !
There is no error now. 

I can find the live data of the **ProbeToTracker** transform in the module "Transforms", however, not sure how to combine it with the Id="TrackedVideoStream", or specifically "**Image_Transducer**".

Could you please let me know how to use it, in order to **build ultrasound volume**? 
I am not sure how to use this transform matrix. Since the Type="VirtualVolumeReconstructor" seems not to be functional without "Reference", I cannot use Id="VolumeReconstructorDevice".

Any tutorial link or ideas about volume generation will help, please.

**Reply by ungi** [reply]
Don't use volume reconstructor in PLUS, because that will further burden that poor computer (Ultrasonix) with more tasks. Use volume reconstructor in Slicer, as shown in this tutorial video: https://youtu.be/2vXeJxYFou4
In the video, we use a Reference. But you can just use "Tracker" instead of "Reference".

**Reply by SenonETS** [reply]
Thank you so much, ungi. I am able to build volume. However, it seems that the volume has to be with cubic shape controlled by ROI. 

Do you know how to record sequences (e.g., 10 scans in a sequence) in 3D, like collecting a 3D video (from the first scan to the 10th) and then show all the scans (10 scans in total) at the same time?

**Reply by lassoan** [reply]
> Do you know how to record sequences (e.g., 10 scans in a sequence) in 3D

This is a separate topic. Please post it as a new question. Thank you!
